dcicutils/project_utils.py

```diff
@@ -0,0 +1,453 @@
+        self.PYPI_NAME = PYPI_NAME or None  # canonicalize '' to None, just in case
+        self.REPO_NAME = REPO_NAME = REPO_NAME or self.repofy(NAME)
+        self.APP_NAME = APP_NAME = APP_NAME or self.appify(REPO_NAME)
+        self.APP_PRETTY_NAME = APP_PRETTY_NAME or self.prettify(APP_NAME)
+        self.PACKAGE_NAME = (PACKAGE_NAME
+                             # This uses ProjectRegistry (not cls) because that part of the protocol is common to all
+                             # classes and subclasses of that class. No need to worry about specialized classes.
+                             # -kmp 19-May-2023
+                             or self.infer_package_name(poetry_data=ProjectRegistry.POETRY_DATA,
+                                                        pypi_name=PYPI_NAME, pyproject_name=PYPROJECT_NAME))
+
+
+class Project:
+    """
+    A class that should be a superclass of all classes registered using ProjectRegistry.register
+
+    All such classes have these names:
+      .PYPROJECT_NAME - The name used in the .register decorator on the class, and in pyproject.toml
+      .NAME - The customary name of the project
+      .REPO_NAME - The name of the repo
+      .PYPI_NAME - The name of the project in pypi (if any), else None
+      .PACKAGE_NAME - The installed name of the project as a package in a venv, for use with pkg_resources
+      .PRETTY_NAME - The pretty name of the package name
+      .APP_NAME - The ame of the project application (see dcicutils.common and the orchestrated app in EnvUtils)
+      .APP_PRETTY_NAME - The pretty name of the project application.
+
+    Some sample usess of pre-defined attributes of a Project that may help motivate the choice of attribute names,
+    though how these get initialized is easier if you're basing your package on C4Project,
+
+    PYPROJECT_NAME | NAME        |REPO_NAME    |PYPI_NAME    |PACKAGE_NAME  |PRETTY_NAME  |APP_NAME  |APP_PRETTY_NAME
+    ---------------+-------------+-------------+-------------+--------------+-------------+----------+----------------
+    dcicsnovault   |snovault     |snovault     |snovault     |snovault      |Snovault     |snovault  |Snovault
+    encoded-core   |encoded-core |encoded-core |encoded-core |encoded-core  |Encoded Core |core      |Core
+    encoded        |cgap-portal  |cgap-portal  |None         |encoded       |CGAP Portal  |cgap      |CGAP
+    encoded        |fourfront    |fourfront    |None         |encoded       |Fourfront    |fourfront |Fourfront
+    encoded        |smaht-portal |smaht-portal |None         |encoded       |SMaHT Portal |smaht     |SMaHT
+
+     The registered name is the one used with the ProjectRegistry.register() decorator.
+    """
+
+    # This contains only the initial spec and is NOT used after initialization. See .identity (and ._identity).
+    # It is declared in this way so that actual declarations of project classes don't show up as syntax errors.
+    IDENTITY: Optional[dict] = None
```

Comment:
`IDENTITY` is a bit of an overloaded term IMO, it may be more clear to name it something else, or at least provide more detail here as it is still not obvious to me

Reply:
How about `NAMES=`?

Reply:
I'm using NAMES =